### PR TITLE
DP-1232 Less events on tokens change

### PR DESCRIPTION
**Pull request by aleksei-berezkin**
This is a member of a triple pull request. Please merge all of them or neither.

Dependent pull requests to be merged synchronously:
~~https://github.com/JetBrains/jetpad-mapper/pull/83~~ https://github.com/JetBrains/jetpad-mapper/pull/84
https://github.com/JetBrains/datapad/pull/36

### CODE REVIEW

**Comment by aleksei-berezkin** [comment]
Changes in last commit: caused by refactoring in mapper.

File: hybrid/src/main/java/jetbrains/jetpad/hybrid/TokenCompleter.java [hunk 1]
```diff
@@ -126,10 +126,15 @@ public Runnable complete(int selectionIndex, Token... tokens) {
...
         TokenListEditor<?> tokenListEditor = getTokenListEditor();
-        tokenListEditor.tokens.remove(index);
-        int i = index;
-        for (Token t : tokens) {
-          tokenListEditor.tokens.add(i++, t);
+
+        if (tokens.length == 0) {
+          tokenListEditor.tokens.remove(index);
+        } else {
+          int targetIndex = index, sourceIndex = 0;
```
Comment (timzam) [inline hybrid/src/main/java/jetbrains/jetpad/hybrid/TokenCompleter.java]: This is not according to the Java code style. There should be one declarations per line.
See http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-141270.html#18761

File: hybrid/src/test/java/jetbrains/jetpad/hybrid/BaseHybridEditorEditingTest.java [hunk 2]
```diff
@@ -1504,6 +1506,64 @@ private void postProcess(List<Token> tokens) {
...
+  private Runnable installUpdatesCounters(final int maxAllowedUpdates) {
+    final PropertyEventHandlers.CountingHandler<Expr> propHandler = new PropertyEventHandlers.CountingHandler<>();
+    final CollectionEventHandlers.CountingHandler<Token> tokensHandler = new CollectionEventHandlers.CountingHandler<>();
+    sync.property().addHandler(propHandler);
+    sync.tokenListEditor().tokens.addHandler(tokensHandler);
+    return new Runnable() {
+      @Override
+      public void run() {
+        assertTrue("Too much property updates: " + propHandler.getCounter(), propHandler.getCounter() <= maxAllowedUpdates);
+        assertTrue("Too much tokens updates: " + tokensHandler.getCounter(), tokensHandler.getCounter() <= maxAllowedUpdates);
```
Comment (timzam) [inline hybrid/src/test/java/jetbrains/jetpad/hybrid/BaseHybridEditorEditingTest.java]: Should be "Too many property updates: ", "Too many tokens updates: ".

File: hybrid/src/test/java/jetbrains/jetpad/hybrid/BaseHybridEditorEditingTest.java [hunk 3]
```diff
@@ -1504,6 +1506,64 @@ private void postProcess(List<Token> tokens) {
...
+    type("1 2");
+    Runnable verifier = installUpdatesCounters(1);
+    backspace();
+    verifier.run();
+  }
+
+  /**
+   * @return Run to verify the count.
+   */
+  private Runnable installUpdatesCounters(final int maxAllowedUpdates) {
```
Comment (timzam) [inline hybrid/src/test/java/jetbrains/jetpad/hybrid/BaseHybridEditorEditingTest.java]: Quite minor but you could rename the method to "installUpdateCounters".

**Comment by timzam** [comment]
Looks good to me. If you can fix the minor issues mentioned above, it would be great.

**Comment by aleksei-berezkin** [comment]
Changes in last commits:
* All recommendations accepted, style fixed;
* Merged from master.

**Pull request merged**